pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -41,6 +46,35 @@ func (s *SharedClients) ExchangeContext(ctx context.Context, key string, conf *d
 	return r, rtt, closer, err
 }
 
+// Called by wrapped TCP connection once downstream connection breaks/is closed as a notfication to close upstream conn.
+func (s *SharedClients) ShutdownClient(key string) {
+	// lock for s.clients access
+	s.lock.Lock()
+	// locate client to re-use if possible.
+	client := s.clients[key]
+	if client == nil {
+		return
+	}
+	s.lock.Unlock()
+
+	client.Lock()
+	defer client.Unlock()
+	client.refcount--
+	if client.refcount == 0 {
```

Comment:
The refCount (reference count) is increased on acquiring the shared client. In the case of UDP: whenever a query arrives, the `key` is derived from the five-tuple. If a shared client with that key is known, a reference to it is acquired (increasing the `refCount`), otherwise a new client is allocated. Once the query has a response, the shared client reference is released (and `refCount` decreases). 
For TCP, after this change, the reference count is not strictly needed, as we lifecycle the client together with the downstream connection - if a downstream connection (from an EP) comes in, we open a corresponding upstream conn, once it is closed, we close the upstream conn.

We need to be careful to do this reference counting primarily for the transparent mode, to avoid attempting to allocate sockets with IP pairs we already have a socket for, since linux will throw an error otherwise.

Note that the method you commented on is specific to TCP. For why this specialisation was necessary, please read the second commit message.

Reply:
Thanks for the detailed explanation. So in case of udp, if multiple queries from same client comes in parallel, then refcount is incremented and same upstream connection is used but if query comes one after other, then new upstream connection is spawned every time as previous connection would be closed. 
A naive question: if downstream is UDP, then upstream will be UDP and there wont be mix like downstream udp and upstream tcp?

Reply:
> So in case of udp, if multiple queries from same client comes in parallel, then refcount is incremented and same upstream connection is used but if query comes one after other, then new upstream connection is spawned every time as previous connection would be closed.

Yes, pretty much. Queries will typically come in in rapid succession due to DNS clients often querying for A and AAAA records immediately. Also, in the case of UDP, it's mildly misleading to speak of "connections" since UDP doesn't have that concept - de facto we mean the socket which is bound to the EP source address/port (!) in transparent mode. 

> if downstream is UDP, then upstream should be UDP and there wont be mix like downstream udp and upstream tcp?

Yes. The proxy will forward the query to upstream using the same protocol as the client used.